Enable os_signposts on all builds

bnham opened the pull request:
#### 860f2d7753d13e4d23202f7f0d0ac1e90112c469
<pre>
Enable os_signposts on all builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=243699">https://bugs.webkit.org/show_bug.cgi?id=243699</a>
rdar://98348637

Reviewed by David Kilzer.

Currently WebKit&apos;s os_signposts are only enabled on AppleInternal builds. The idea was that
sensitive content might be included in the signposts, so we restricted emitting them to only
internal builds.  But this is the wrong way of achieving that goal, as we can use the standard
logging privacy specifiers to control content that should be redacted from signposts. This also
makes it hard to enable the signposts on perf test bots that emulate standard OS builds.

Another issue is that there are two ways to emit tracing events: kdebug tracepoints and os_signpost.
It&apos;s confusing that there are two ways to do almost the same thing. The current guidance from our
system teams is to use signposts for new trace points, since they&apos;re more flexible (e.g. allow for
logging free-form strings). But using signposts in WebKit requires multiple manual steps which this
patch aims to address.

This patch tries to address these issues by:

 1. Removing the AppleInternal restriction for os_signposts. Redaction will be done via standard
    logging privacy specifiers like `%{private}`.
 2. Adding a build phase to WebKit&apos;s Xcode project that installs a plist to /S/L/Preferences/Logging
    that enables persistence of WebKit signposts by default. Many other subsystems already do
    this. This makes it so that develoeprs don&apos;t have to manually run a step to install such a
    profile on their own to get the signposts to show up reliably in our tracing tool.
 3. By default, make `WTF{Emit,Begin,End}Signpost` only emit the signpost when kdebug tracing is
    enabled. This reduces the chances that we pollute the persistent log with signposts when they&apos;re
    not useful. It also matches the behavior of `kdebug_trace` (which is used by `WTF::tracePoint`).

With these changes, we should be able to provide simplified guidance on how to use signposts to
other WebKit developers.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp
(signpostStart):
(signpostStop):
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled): Deleted.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Resources/LoggingPreferences.plist: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled): Deleted.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Resources/Signposts/LoggingPreferences.plist: Added.
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist: Copied from Tools/Tracing/SystemTracePoints.plist.
* Source/WebKit/Scripts/copy-signpost-plists.sh: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/Tracing/SystemTracePoints.plist: Changed to be a symlink to Source/WebKit/Resources/Signposts/SystemTracePoints.plist.

Canonical link: <a href="https://commits.webkit.org/253387@main">https://commits.webkit.org/253387@main</a>
</pre>

ddkilzer commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -16663,6 +16666,26 @@
+			);
+			runOnlyForDeploymentPostprocessing = 1;
+			shellPath = /bin/sh;
+			shellScript = "set -e\n\nif [[ \"${WK_PLATFORM_NAME}\" == maccatalyst || \"${WK_PLATFORM_NAME}\" == iosmac ]]; then\n    exit 0\nfi\n\ncp \"${SCRIPT_INPUT_FILE_0}\" \"${SCRIPT_OUTPUT_FILE_0}\"\nplutil -convert binary1 \"${SCRIPT_OUTPUT_FILE_0}\"\n";
```
Why no signposts for MacCatalyst apps?

If we can use signposts on MacCatalyst builds, can we use a plain "copy file" build phase instead of a custom script here?

bnham replied:
The build phase is to copy a preference plist to `/System/Library/Preferences/Logging/Subsystems/com.apple.WebKit.plist`. This changes all signposts in the `com.apple.WebKit` subsystem to be persisted by default (rather than in-memory only). Lots of subsystems do this, the reason being that in-memory signpost storage makes it so that you see only an arbitrary number of signposts in your artrace, depending on the state of the in-memory log buffer at the time of the trace. That makes it harder to get started with tracing.

We skip copying the file for `maccatalyst` and `iosmac` because as far as I understand, `maccatalyst` and `iosmac` do not have their own logd. Instead they just use the logd that's running as part of the base macOS system.

ddkilzer replied:
Makes sense.  r=me

emw-apple replied:
I would prefer a native Copy Files build phase, with EXCLUDED_SOURCE_FILE_NAMES set in WebKit.xcconfig to avoid copying LoggingPreferences.plist for catalyst builds. It avoids us having to ensure the input/output paths are declared correctly, and is a bit less overhead.

bnham replied:
@emw-apple Is there a way to rename a file in a Copy Files build phase? I'm copying these files:

```
$(SRCROOT)/Resources/Signposts/LoggingPreferences.plist
$(SRCROOT)/Resources/Signposts/SystemTracePoints.plist
```

To these destinations:

```
$(DSTROOT)/System/Library/Preferences/Logging/Subsystems/com.apple.WebKit.plist
$(DSTROOT)/AppleInternal/Library/Ariadne/Plists/WebKit.plist
```

I could rename the source filenames to match the destination filenames, but the filenames `com.apple.WebKit.plist` and `WebKit.plist` are awfully generic, and seem like they would pollute Xcode's Open Quickly, etc.

emw-apple replied:
You could do it using a build rule (and APPLY_RULES_IN_COPY_FILES=YES, which WebKit should already have enabled), though you'd have to write out a shell script to do the copying, at which point you might as well use the original script phase 😄

That said, I think it might actually be better to keep the names the same, since it would allow someone to see the file in build products (or installed in a root) and more easily figure out where that file comes from in the project. The `Source/WebKit/Resources/Signposts` path (which Xcode shows in Open Quickly) makes up for some of that lost context.

bnham replied:
I don't anticipate that these files will be changed often (or at all). Given that, I think for now I want to err on the side of not polluting people's autocomplete for `WebKit` in Xcode. But if you feel really strongly then let me know and I will file a follow up patch to replace this build phase with a copy files script.

emw-apple replied:
I don't feel strongly, LGTM :)

ddkilzer reviewed (approved):
r=me, but please consider comment about the build phase script.

bnham committed:
Enable os_signposts on all builds
https://bugs.webkit.org/show_bug.cgi?id=243699
rdar://98348637

Reviewed by David Kilzer.

Currently WebKit's os_signposts are only enabled on AppleInternal builds. The idea was that
sensitive content might be included in the signposts, so we restricted emitting them to only
internal builds.  But this is the wrong way of achieving that goal, as we can use the standard
logging privacy specifiers to control content that should be redacted from signposts. This also
makes it hard to enable the signposts on perf test bots that emulate standard OS builds.

Another issue is that there are two ways to emit tracing events: kdebug tracepoints and os_signpost.
It's confusing that there are two ways to do almost the same thing. The current guidance from our
system teams is to use signposts for new trace points, since they're more flexible (e.g. allow for
logging free-form strings). But using signposts in WebKit requires multiple manual steps which this
patch aims to address.

This patch tries to address these issues by:

 1. Removing the AppleInternal restriction for os_signposts. Redaction will be done via standard
    logging privacy specifiers like `%{private}`.
 2. Adding a build phase to WebKit's Xcode project that installs a plist to /S/L/Preferences/Logging
    that enables persistence of WebKit signposts by default. Many other subsystems already do
    this. This makes it so that develoeprs don't have to manually run a step to install such a
    profile on their own to get the signposts to show up reliably in our tracing tool.
 3. By default, make `WTF{Emit,Begin,End}Signpost` only emit the signpost when kdebug tracing is
    enabled. This reduces the chances that we pollute the persistent log with signposts when they're
    not useful. It also matches the behavior of `kdebug_trace` (which is used by `WTF::tracePoint`).

With these changes, we should be able to provide simplified guidance on how to use signposts to
other WebKit developers.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp
(signpostStart):
(signpostStop):
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled): Deleted.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Resources/LoggingPreferences.plist: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFSignpostsEnabled): Deleted.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Resources/Signposts/LoggingPreferences.plist: Added.
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist: Copied from Tools/Tracing/SystemTracePoints.plist.
* Source/WebKit/Scripts/copy-signpost-plists.sh: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/Tracing/SystemTracePoints.plist: Changed to be a symlink to Source/WebKit/Resources/Signposts/SystemTracePoints.plist.

Canonical link: https://commits.webkit.org/253387@main
